python/ray/util/sgd/torch/torch_trainer.py

```diff
@@ -647,23 +646,16 @@ def default_resource_request(cls, config):
                     "num_cpus_per_worker", kwargs.get("num_cpus_per_worker",
                                                       1))
                 use_gpu = config.get("use_gpu", kwargs.get("use_gpu"))
-                use_local = config.get("use_local",
-                                       kwargs.get("use_local", False))
 
-                if use_local:
-                    remote_worker_count = num_workers - 1
-                    local_cpus = 1
-                    local_gpus = int(use_gpu)
-                else:
-                    remote_worker_count = num_workers
-                    local_cpus = 0
-                    local_gpus = 0
-
-                return Resources(
-                    cpu=int(local_cpus * num_cpus_per_worker),
-                    gpu=int(local_gpus),
-                    extra_cpu=int(remote_worker_count * num_cpus_per_worker),
-                    extra_gpu=int(int(use_gpu) * remote_worker_count))
+                bundles = [
+                    # Worker bundles
+                    {
+                        "CPU": num_cpus_per_worker,
+                        "GPU": int(use_gpu)
+                    }
+                ] * num_workers
+
+                return PlacementGroupFactory(bundles, strategy="PACK")
```

Comment:
Isn't the first bundle for the `Trainable` itself, so it should not request any GPUs or `num_cpus_per_worker` CPUs? Is it possible for the first bundle to just be 0 CPUs?

Reply:
No, zero-resource bundles raise an error on the ray core side. Previously I added a 1 CPU bundle if local_mode is False, but this would then require more resources than before.

Reply:
I see. Then having the 1 CPU bundle is `use_local` is False is what we want right? Otherwise the Trainable itself will request too many resources, and there will be less bundles than num_workers?

Reply:
Will it though? Previously we didn't request a CPU for the driver process and `num_cpus=0` for `local_mode=False` - e.g. the number of requested resources should be the same here. 

The difference for the `local_mode` is thus just here:

```
        if self.use_local:
            self.worker_group = LocalWorkerGroup(**worker_args)
        else:
            self.worker_group = RemoteWorkerGroup(**worker_args)
 ```
 
 The local trainer of the LocalWorkerGroup will live in the same bundle as the driver, but that shouldn't be a problem as long as the driver doesn't occupy any resource. Or am I missing something here?

Reply:
My understanding is that when using placement groups with Tune, the first bundle is for the Trainable and any subsequent bundles will be used for additional actors that are launched by the trainable. When using RaySGD with Tune, the driver creates the Tune Trainables, and each Tune Trainable instantiates a `TorchTrainer` which creates `num_workers` actors. In the case where `use_local` is True, the first training worker will be on the same process as the Trainable, so the Trainable needs enough resources to launch a worker and the current code works for that case. However when `use_local` is False, all of the workers are remote actors, so the Trainable itself should only request 1 CPU any not any GPUs. So I believe the bundles for the `use_local` False case should be `[{"CPU": 1}] + [{"CPU": cpus_per_worker, "GPU": int(use_gpu)}] * num_workers`

Reply:
Ok so generally this should still work, as the `use_local` just switches between calling the first worker in a `.remote()` call or adding it directly in the driver.

I.e. this code works nicely:

```
import ray
from ray.util import placement_group

ray.init(num_cpus=4)


@ray.remote(num_cpus=4)
def worker_task():
    print(ray.get_resource_ids())
    return 1


@ray.remote
def driver_task():
    print(ray.get(worker_task.remote()))
    return 2


pg = placement_group([{"CPU": 4}])
print(ray.get(driver_task.options(placement_group=pg, num_cpus=0).remote()))
```

However, you are right: In the current placement group implementation the driver process will always occupy the full resources of the first bundle, so there wouldn't be resources left for the remote trainable.

Long story short: Thanks for catching this! I adjusted the code :-)